doc/tests-updates.md

```diff
@@ -0,0 +1,381 @@
+Don't know how to implement it yet.
+
+- Debug file. - Filename to write debug information.
+
+Server is stopped by `TERM` signal from the framework.
+
+Script contains queue of expected requests `server_data.requests[]`. If all
+requests was received, but a new request is received, then client is working in
+repeated mode, thus `conn_data.expected_request_num` must be reset to `0`.
+
+`def gen_response_func(server_data, conn_data, request)` function is used to
+give a response for the client. Returns `response` or `None`. The received
+request is compared with `server_data.requests[T-Req-Id]`
+but note, that `T-Client-Id`, `T-Conn-Id`, `T-Req-Id` values may be missed in the
+expected request.
+
+The generated response **must** contain headers copied from the request:
+`T-Client-Id`, `T-Conn-Id`, `T-Req-Id`. `T-Srv-Id` and `T-Srv-Conn-Id` must
+also be generated.
+
+If the server need to close the connection in response to request, it sets
+`conn_data.close` variable. The variable `conn_data.expect_close` can also be
+set if generated response is invalid and Tempesta will reset the connection.
+
+
+### Framework
+
+The framework now is defined not as list of actions but as HTTP flows on client
+and server. This allows to replace Tempesta by other HTTP ADC software and easily
+compare behavior between them.
+
+This also allows to create multiple tests from the same description:
+
+- Message integrity, Invalid message processing, Cache tests: single client,
+single client connection.
+
+- Scheduler, Workload, Request-response chain integrity tests: multiple clients
+from many source interfaces, multiple concurrent connections for each client.
+
+- Fuzzing tests: Workload tests, but every generated messages are split into
+different number of skbs.
+
+When reduced number of tests is executed, only first group of tests is executed,
+the full set is for first and second groups. All groups are executed in background
+stability tests.
+
+At the first sight Cache tests can't be executed multiple times.
+But if `URI` and/or `Host:` mutated between different client connections and
+repeats it becomes possible. Thus Cache Workload testing should be fine.
+
+Fault injections tests are very similar to Message integrity tests, but a few
+prior actions to inject errors are required before starting the client.
+
+
+#### Prepare to start
+
+Copy functions defined in `clients` and `server` array of test class to separate
+file to start clients and servers as distinct processes. Probably other python
+interpreters can be used for that. `Pypi` is extremely fast in comparing to
+standard interpreter. This will allow pretty good RPS in work load tests.
+
+The framework must create all required interfaces before start.
+
+POSIX barrier is used to synchronize clients and servers. Clients shouldn't
+start generating requests before all server connections are established.
+
+
+## TBD
+
+TLS tests - ?
+
+Low level TCP testing - ?
```

Comment:
With the recent TLS changes affecting TCP transmission process we have to test TCP operation (@i-rinat has questions regarding TCP segment sized generated by Tempesta). The question will be even harder for HTTP/3. Probably ScaPy isn't a good option for this, because it'd be too tricky to generate TCP flow by hands. Probably iptables mangling and/or eBPF with usual TCP options like Nagle will suite the best. We already have something in `helpers/analyzer.py` and `helpers/flaky.py`.